src/main/java/com/iexec/worker/compute/ComputeExitCauseService.java

```diff
@@ -16,78 +16,81 @@
-    }
-
-    /**
-     * Get exit cause for pre-compute or post-compute enclave.
-     *
-     * @param chainTaskId task ID
-     * @return exit cause
-     */
-    ReplicateStatusCause getReplicateStatusCause(ComputeStage computeStage,
-                                                         String chainTaskId) {
-        return exitCauseMap.get(buildKey(computeStage, chainTaskId));
-    }
 
-    /**
-     * Get pre-compute exit cause.
-     *
-     * @param chainTaskId task ID
-     * @return exit cause
-     */
-    public ReplicateStatusCause getPreComputeExitCauseAndPrune(String chainTaskId) {
-        ComputeStage stage = ComputeStage.PRE;
-        ReplicateStatusCause cause = getReplicateStatusCause(stage, chainTaskId);
-        pruneExitCause(stage, chainTaskId);
-        return cause != null ? cause : ReplicateStatusCause.PRE_COMPUTE_FAILED_UNKNOWN_ISSUE;
+        final String key = buildKey(computeStage, chainTaskId);
+        exitCauseMap.compute(key, (k, existingCauses) -> {
+            if (existingCauses == null) {
+                log.info("Added bulk exit causes [computeStage:{}, chainTaskId:{}, causeCount:{}]",
+                        computeStage, chainTaskId, causes.size());
+                return List.copyOf(causes);
+            } else {
+                log.info("Appended bulk exit causes to existing list [computeStage:{}, chainTaskId:{}, newCauseCount:{}, totalCauseCount:{}]",
+                        computeStage, chainTaskId, causes.size(), existingCauses.size() + causes.size());
+                final List<ReplicateStatusCause> combinedCauses = new ArrayList<>(existingCauses);
+                combinedCauses.addAll(causes);
+                return combinedCauses;
+            }
+        });
+        return true;
     }
 
     /**
-     * Get post-compute exit cause.
+     * Get exit causes for a specific compute stage and prune them.
+     * Returns default unknown issue cause when no specific causes are set.
      *
-     * @param chainTaskId task ID
-     * @return exit cause
+     * @param computeStage compute stage
+     * @param chainTaskId  task ID
+     * @return list of exit causes, or default unknown issue if not found
      */
-    public ReplicateStatusCause getPostComputeExitCauseAndPrune(String chainTaskId) {
-        ComputeStage stage = ComputeStage.POST;
-        ReplicateStatusCause cause = getReplicateStatusCause(stage, chainTaskId);
-        pruneExitCause(stage, chainTaskId);
-        return cause != null ? cause : ReplicateStatusCause.POST_COMPUTE_FAILED_UNKNOWN_ISSUE;
+    public List<ReplicateStatusCause> getExitCausesAndPruneForGivenComputeStage(ComputeStage computeStage, String chainTaskId) {
+        final String key = buildKey(computeStage, chainTaskId);
+        final List<ReplicateStatusCause> causes = exitCauseMap.remove(key);
+        if (causes != null) {
+            log.debug("Retrieved and pruned bulk exit causes [computeStage:{}, chainTaskId:{}, causeCount:{}]",
```

Comment:
I advocate against using **bulk** here.
We are allowing to manipulate a list of issues.
It has nothing to do wit bulk processing, even if it is this project which creates the need of an improvement here:
```suggestion
            log.debug("Retrieved and pruned exit causes [computeStage:{}, chainTaskId:{}, causeCount:{}]",
```

Reply:
Resolved in [3b092b9](https://github.com/iExecBlockchainComputing/iexec-worker/pull/653/commits/3b092b93e83c3ded87868869aa4c2e8929b7a54b)